### PR TITLE
Enable Native Wayland Mode

### PR DESCRIPTION
**Pull request by kaimast**
Hi,

This pull request merges the changes from #188 and #182, as well as ensuring Zoom runs in native Wayland mode when possible.

Note that screensharing currently still does not work on Wayland or XWayland. I think they might have moved to a different API when migrating to a newer version of Qt, but not sure.
However, this branch behaves the same as master on X11.

-Kai

### CODE REVIEW

**Comment by gncs** [comment]
Looks good to me. What's left before we can merge this?

File: us.zoom.Zoom.json
Comment (GeorgesStavracas) [inline us.zoom.Zoom.json]: Here you add this, and then immediately removes it in the next commit. I'd rather not add it here, and remove the next commit.

File: zoom.sh
Comment (GeorgesStavracas) [inline zoom.sh:7]: I think only `wayland` is the appropriate one, even if `wayland-egl` works just fine.

**Review by GeorgesStavracas — CHANGES_REQUESTED** [review]

File: us.zoom.Zoom.json
Comment (Yajo) [inline us.zoom.Zoom.json:23]: This seems a bug:

```suggestion
        "--talk-name=org.freedesktop.ScreenSaver",
        "--talk-name=org.freedesktop.PowerManagement.Inhibit",
```

**Comment by vchernin** [comment]
bot, build

**Comment by vchernin** [comment]
Branch seems to work well on native Wayland. However, it does cause a non-GNOME style header bar. Presumably QT's default CSD on GNOME Wayland isn't the GNOME style.

Perhaps there's a way to change that, but I wouldn't personally consider it a blocking issue. There still is a close button in the top right.

File: us.zoom.Zoom.json
Comment (kaimast) [reply us.zoom.Zoom.json]: Thanks! I fixed that just now.

File: us.zoom.Zoom.json
Comment (kaimast) [reply us.zoom.Zoom.json]: Yeah... I was playing around with a bunch of different configurations initially. Ideally we can just squash my commits if this ever gets merged?

**Comment by kaimast** [comment]
@vchernin I think that theming will be a general problem with CSD on GNOME.

Afaik, Zoom is still using Qt 5.12. Maybe a more recent version has better integration with GTK/GNOME, but there is not much we can do about this right now.

File: zoom.sh
Comment (kaimast) [reply zoom.sh]: Thanks! I changed that.

**Comment by vchernin** [comment]
I've managed to reproduce an issue with this branch. I've never experienced anything like it when Zoom wasn't in Wayland native mode.

To reproduce:
1. Join a zoom meeting that is being recorded.
2. Try to accept the recording acknowledgement dialog.
3. I am never able to get past this dialog, every time I attempt to join zoom crashes right after displaying the dialog.

Worst of all this issue only occurs on my 1440p 200% scaling laptop. It does not occur when connected to my 4k 200% scaling monitor. I have no idea why, I think it's very likely some strange QT edge case.

I might try to do some further debugging to narrow down the cause, but for now I think this shouldn't be merged. I wasn't even trying to find issues yet I found one that completly prevents me from joining a meeting.

**Comment by kaimast** [comment]
I updated this pull request to use the Kde runtime. On my system this results in a proper GTK theme being used for the title bars!

Also, I do agree that we should not merge this until it is crash-free. We probably have to wait for the next major Zoom update (at least), when they will hopefully move to a newer version of Qt with better Wayland integration.

**Comment by vchernin** [comment]
A couple issues I noticed with 51081:
It appears to depend on both `org.freedesktop.Platform.html5-codecs//18.08` and `org.freedesktop.Platform.VAAPI.Intel//18.08`. This doesn't inhibit the app's functionality but it probably shouldn't need these old runtimes. This didn't occur in older builds. I wonder if this is caused by the switch to the kde 5.12 runtime.

The first time the app opens this message pops up with broken antialiasing. Doing as suggested seems to fix the issue (set `enableWaylandShare=true`), even on reinstall. Fonts are fine everywhere else I've looked in the app. 
no antialiasing pop up warning

It appears that installing the flatpak system wide (i.e. without --user) doesn't solve this first time, the same fix is still needed, even though I have previously tried to set that value system wide. Also fixing the config file doesn't actually improve screen sharing behaviour, all it does is get rid of the prompt. On my device it's still broken as ever due to #22. 

When I'm signed in to a zoom account I am only sometimes able to click the profile button. Sometimes if click it several times quickly it works? (it works normally in the stable branch).
profile button (bottom right)

On the bright side I tried to reproduce [that recording issue](https://github.com/flathub/us.zoom.Zoom/pull/255#issuecomment-852730531) and it doesn't seem to occur now. Progress!

**Comment by jeremy-lilly** [comment]
Just in case it's helpful, I can confirm that this worked for me using the proprietary Nvidia drivers (470.63.01) in Arch under GNOME Wayland. I had the same issue described above where Zoom was not using my system theme.

**Comment by vchernin** [comment]
@kaimast how feasible is it do you think to make native Wayland mode opt in in the stable build? What I mean is by default Zoom stable will run in Xwayland mode, but if the user enables the Wayland socket in something like Flatseal Zoom will run in Wayland natively. So in other words only users who seek out the Wayland socket would get native Wayland support.

Since there are increasingly more reasons why users would want to run in Wayland natively, I think making this more accessible would be quite neat. 

See https://github.com/flathub/com.slack.Slack/pull/139 for inspiration

**Comment by dac73** [comment]
bot, build

File: zoom.sh
Comment (vchernin) [inline zoom.sh:6]:
```suggestion
if [ -z "$DISPLAY" ] && [ -n "$WAYLAND_DISPLAY" ]; then
```

True if actively running in Wayland and Wayland Flatpak socket is enabled. Together with the above change (don't pass Wayland socket by default) means Zoom will run in Xwayland by default, and if someone enables the Wayland Flatpak socket it will run in Native Wayland mode.

Inspired by https://github.com/flathub/com.slack.Slack/pull/139

File: us.zoom.Zoom.json
Comment (vchernin) [inline us.zoom.Zoom.json:13]:
```suggestion

```

With below change.

**Review by vchernin — CHANGES_REQUESTED** [review]
With these changes I think this is mergeable into stable.

File: zoom.sh
Comment (vchernin) [inline zoom.sh:25]:
```suggestion

if [ -z "$DISPLAY" ] && [ -n "$WAYLAND_DISPLAY" ]; then
    QPA=wayland
fi
```

True if actively running in Wayland and Wayland Flatpak socket is enabled. Together with the below change (don't pass Wayland socket by default) means Zoom will run in Xwayland by default, and if someone enables the Wayland Flatpak socket (and disables X11 socket/enables fallback X11) Zoom will run in Native Wayland mode.

File: zoom.sh
Comment (vchernin) [inline zoom.sh:7]:
```suggestion

```

File: zoom.sh
Comment (vchernin) [inline zoom.sh:23]:
```suggestion
```

File: zoom.sh
Comment (vchernin) [inline zoom.sh:4]:
```suggestion
QPA=xcb
```

**Comment by kaimast** [comment]
Thanks vchernin. Will try to do this tonight.

**Comment by kaimast** [comment]
This looks good now. I added a README file that explains you have to enable the Wayland socket AND disable the x11 socket as well. That latter was not clear to me initially and might be confusing to some other folks as well.

**Comment by vchernin** [comment]
For what its worth you could also enable fallback-X11 + Wayland sockets. But yeah it can be weird to wrap your head around.

File: README.md
Comment (Yajo) [inline README.md:8]:
```suggestion
This can be done, for example, through Flatseal or through the command line.
```

**Review by Yajo — CHANGES_REQUESTED** [review]

**Comment by Yajo** [comment]
> you could also enable fallback-X11 + Wayland sockets

Maybe a nice default?

**Comment by kaimast** [comment]
> > you could also enable fallback-X11 + Wayland sockets
> 
> Maybe a nice default?

I think the whole point of the recent change is to disable Wayland by default.
It does not seem like screensharing support will be there any time soon, so leaving it opt-in is the best way to go.

**Comment by vchernin** [comment]
>  It does not seem like screensharing support will be there any time soon, so leaving it opt-in is the best way to go.

Indeed. I think the main thing to test here now is if using KDE 5.15 over freedesktop 20.08 introduces any bugs, I kinda doubt it though.

**Comment by kaimast** [comment]
FWIW I've been using this branch for a while now to teach a class and have not encountered any major issues.

**Comment by vchernin** [comment]
> FWIW I've been using this branch for a while now to teach a class and have not encountered any major issues.

On Wayland natively? If so that is good to hear, and it likely means X11 will be fine too.

**Comment by kaimast** [comment]
> > FWIW I've been using this branch for a while now to teach a class and have not encountered any major issues.
> 
> On Wayland natively? If so that is good to hear, and it likely means X11 will be fine too.

No, I have mostly been using X11 because I need to share slides with my students :laughing:

**Comment by Yajo** [comment]
> I think the whole point of the recent change is to disable Wayland by default.
> It does not seem like screensharing support will be there any time soon, so leaving it opt-in is the best way to go.

Sorry, the PR title confused me.

So, basically, wayland + zoom = no screen share. No matter what you do. Is that it?

**Comment by vchernin** [comment]
> So, basically, wayland + zoom = no screen share. No matter what you do. Is that it?

From #22 it seems that way... it is not a great situation. It worked in the past more reliably it seems. Maybe it works better on non Silverblue distros.

**Comment by Iolaum** [comment]
If screen sharing doesn't reliably work it is better to not merge this PR yet IMO. As it is, screen sharing works and it 'd be better to keep it that way.

**Comment by lolsniperftw** [comment]
is it still being looked at?

**Comment by kaimast** [comment]
> is it still being looked at?

I'm not sure if this should be merged yet. Wayland support is still pretty bad. Maybe at some point Zoom updates to Qt6 and things will get better? 

You can always set `QT_QPA_PLATFORM=wayland` in Flatseal if you want to use Zoom on Wayland instead.

**Comment by czhang03** [comment]
Just a note, zoom can screen share via pipwire just fine on wayland, without adding additional socket https://wiki.archlinux.org/title/Zoom_Meetings#Wayland. Might want to mention that in the readme.

**Comment by kaimast** [comment]
Closing this as I do not think it works anymore. At least on my end, QT_QPA_PLATFORM seems to be ignored with Zoom 6.0.

**Pull request closed without merging**
